crates/telio-traversal/src/session_keeper.rs

```diff
@@ -111,113 +138,202 @@ impl SessionKeeper {
+                                            IpAddr::V6(_) => &c.pingers.pinger_client_v6,
+                                        };
+
+                                        let _ = secondary_client
+                                            .pinger(second, PingIdentifier(rand::random()))
+                                            .await
+                                            .send_ping(PingSequence(0), &[0; PING_PAYLOAD_SIZE])
+                                            .await?;
+                                    }
+                                }
+                                Ok(())
+                            })
+                        }),
+                    ))
+                })
+                .await
+                .map_err(Error::Task)?
+                .map_err(Error::RepeatedActionError)
+                .map(|_| ());
+
+                res?;
+            }
+            AddNodeParameters::Batching {
+                public_key,
+                target,
+                interval,
+                threshold,
+            } => {
+                let dual_target = DualTarget::new(target).map_err(Error::DualTargetError)?;
+
+                let rx = self
+                    .batcher
+                    .add(&public_key.to_string(), interval, threshold)
+                    .await?;
+
+                task_exec!(&self.task, async move |s| {
+                    s.batched_actions.insert(
+                        public_key,
+                        PingData {
+                            trigger_rx: rx,
+                            targets: dual_target,
+                        },
+                    );
+
+                    Ok(())
+                })
+                .await?;
+            }
+        }
 
-    async fn update_interval(&self, public_key: &PublicKey, interval: Duration) -> Result<()> {
-        let public_key = *public_key;
-        task_exec!(&self.task, async move |s| {
-            let _ = s.keepalive_actions.update_interval(&public_key, interval);
-            Ok(())
-        })
-        .await
-        .map_err(Error::Task)
+        Ok(())
     }
 
-    async fn remove_node(&self, public_key: &PublicKey) -> Result<()> {
-        // FIXME: error handling with task_exec! seems to suck a lot. Need to fix that.
-        let public_key = *public_key;
-        task_exec!(&self.task, async move |s| {
-            let _ = s.keepalive_actions.remove_action(&public_key);
-            Ok(())
-        })
-        .await
-        .map_err(Error::Task)
+    async fn remove_node<'a>(&self, params: RemoveNodeParameters<'a>) -> Result<()> {
+        match params {
+            RemoveNodeParameters::NonBatching { public_key } => {
+                let public_key = *public_key;
+                let _ = self.batcher.remove(&public_key.to_string()).await;
+
+                task_exec!(&self.task, async move |s| {
+                    if s.nonbatched_actions.contains_action(&public_key) {
```

Comment:
nit: this check is probably not needed

Reply:
You're right, it doesn't check for any errors thus removing this is the way to go, fixed it